### PR TITLE
Fix race conditions for Azure loadbalancer and route updates

### LINKED ISSUE #75582: Race condition of securityGroup/routeTable/loadBalancer setup in Azure cloud provider
**What happened**:

Azure cloud provider configures securityGroup, routeTable and loadBalancer by invoking ARM APIs. But on cluster initialization stage, it may update them even when the create operation from provisioning tools doesn't finish. Because the later PUT would cancel a previous ongoing one, this would introduce troubles for provisioning tools.

We should check the status of securityGroup and routeTable, and only do updates for them when they are in state `Succeeded | Failed | Canceled`.

Also for a running cluster, other components may also update those resources. So ETAG for each resource should be honored when the cloud provider updates them.

**What you expected to happen**:

Cloud provider shouldn't cancel operations from provisioning tools.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig azure
/kind bug
/priority important-soon
/milestone v1.15

### PR DESCRIPTION
**Pull request by feiskyer**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/kind bug
/sig azure
/priority important-soon

**What this PR does / why we need it**:

Azure cloud provider configures routeTable and loadBalancer by invoking ARM APIs. But on cluster initialization stage, it may update them even when the create operation from provisioning tools doesn't finish. Because the later PUT would cancel a previous ongoing one, this would introduce troubles for provisioning tools.

This PR fixes race conditions for them by adding ETAG in PUT requests, hence cloud provider won't override the concurrent updates from other components.

**Which issue(s) this PR fixes**:

Fixes #75582

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Fix race conditions for Azure loadbalancer and route updates.
```

### CODE REVIEW

**Comment by feiskyer** [comment]
Doing more validations for race conditions.

/hold
/test pull-kubernetes-e2e-aks-engine-azure

**Comment by feiskyer** [comment]
/hold cancel

Ready for review now. @andyzhangx PTAL

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_backoff.go [hunk 1]
```diff
@@ -454,8 +468,19 @@ func (az *Cloud) createOrUpdateRouteTableWithRetry(routeTable network.RouteTable
...
+			az.rtCache.Delete(*routeTable.Name)
+			return done, nil
+		}
+
+		// Invalidate the cache and abort backoff because ETAG precondition mismatch.
+		if resp != nil && resp.StatusCode == http.StatusPreconditionFailed {
+			az.rtCache.Delete(*routeTable.Name)
+			return true, err
+		}
+		return done, retryError
```
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_backoff.go]: if it's <false, error> , then why keep the cache? what about always clean the cache?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_backoff.go [hunk 2]
```diff
@@ -479,9 +507,20 @@ func (az *Cloud) createOrUpdateRouteWithRetry(route network.Route) error {
...
+			az.rtCache.Delete(az.RouteTableName)
+			return done, nil
+		}
+
+		// Invalidate the cache and abort backoff because ETAG precondition mismatch.
+		if resp != nil && resp.StatusCode == http.StatusPreconditionFailed {
+			az.rtCache.Delete(az.RouteTableName)
+			return true, err
+		}
+		return done, retryError
```
Comment (andyzhangx) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_backoff.go]: what about always clean the cache for error condition?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_backoff.go (hunk 1 above)
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_backoff.go]: for those cases, cloud provider should backoff retry with exactly same data. re-get the data from ARM API is not needed, as the ETAG is not conflicted.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_backoff.go (hunk 2 above)
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_backoff.go]: same with above, re-get the data is not needed because etag is still same.

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Pull request merged**